axiom/tokens_integration_test.go

```diff
@@ -0,0 +1,91 @@
+//go:build integration
+
+package axiom_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/suite"
+
+	"github.com/axiomhq/axiom-go/axiom"
+)
+
+// TokensTestSuite tests all methods of the Axiom Tokens API against a
+// live deployment.
+type TokensTestSuite struct {
+	IntegrationTestSuite
+
+	apiToken *axiom.APIToken
+}
+
+func TestTokensTestSuite(t *testing.T) {
+	suite.Run(t, new(TokensTestSuite))
+}
+
+func (s *TokensTestSuite) SetupSuite() {
+	s.IntegrationTestSuite.SetupSuite()
+}
+
+func (s *TokensTestSuite) TearDownSuite() {
+	// Teardown routines use their own context to avoid not being run at all
+	// when the suite gets cancelled or times out.
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+
+	err := s.client.Tokens.Delete(ctx, s.apiToken.ID)
+	s.NoError(err)
+
+	s.IntegrationTestSuite.TearDownSuite()
+}
+
+func (s *TokensTestSuite) Test() {
+	createdToken, err := s.client.Tokens.Create(s.suiteCtx, axiom.CreateTokenRequest{
+		Name:      "Test token",
+		ExpiresAt: time.Now().Add(24 * time.Hour),
+		DatasetCapabilities: map[string]axiom.DatasetCapabilities{
+			"*": {Ingest: []string{"create"}}},
+		OrganisationCapabilities: axiom.OrganisationCapabilities{
+			Users: []string{"create", "read", "update", "delete"},
+		}})
+	s.Require().NoError(err)
+	s.Require().NotNil(createdToken)
+
+	s.apiToken = createdToken.AsAPIToken()
+
+	// Get the token and make sure it matches what we have updated it to.
+	token, err := s.client.Tokens.Get(s.ctx, s.apiToken.ID)
+	s.Require().NoError(err)
+	s.Require().NotNil(token)
+
+	s.Equal(s.apiToken, token)
+
+	// List all tokens and make sure the created token is part of that
+	// list.
+	tokens, err := s.client.Tokens.List(s.ctx)
+	s.Require().NoError(err)
+	s.Require().NotEmpty(tokens)
+
+	s.Contains(tokens, s.apiToken)
+
+	// Regenerate the token and make sure the new token is part of the list.
+	regeneratedToken, err := s.client.Tokens.Regenerate(s.ctx, s.apiToken.ID, axiom.RegenerateTokenRequest{
+		ExistingTokenExpiresAt: time.Now().Add(-time.Second),
```

Comment:
Why the negative offset?

Reply:
No reason but to ensure that the newly existing token is deleted

Reply:
Ah, got it. So the server will basically list by doing `SELECT * FROM api_tokens WHERE expiry >= NOW()` (pseudo code, obviously 😄)? So you're trying to make sure it is definitely excluded, even if not yet purged from the table?

I'm not sure it matters, tbh. The latency should be high enough to prevent this from happening, e.g. `List/Get` will happen far enough after `Regenerate`.

Have you observed it failing? If not, either delete this **or** add a `HINT(topper codes): ...` one-liner, please.

Reply:
I've not observed it so will remove it